### PR TITLE
Check that count_mat is of integer type

### PR DESCRIPTION
**Pull request by igordot**
Non-integer matrices will eventually cause an error

### CODE REVIEW

**Comment by evanbiederstedt** [comment]
Hi @igordot 

Could you provide a few more details here? 

The error exists if the matrix is numeric in type as well?

File: R/utils.R [hunk 1]
```diff
@@ -1634,7 +1634,10 @@ log_err = function(msg) {
 #' @keywords internal
 check_matrix = function(count_mat) {
     if ('matrix' %in% class(count_mat)) {
-        count_mat <- as(Matrix(count_mat, sparse=TRUE), "dgCMatrix")
+        if (typeof(count_mat) != 'integer') {
+            log_err("count_mat is not of integer type")
```
Comment (evanbiederstedt) [inline R/utils.R:1638]: I'd prefer a more informative error. 

The logic from Teng is to check if it's a matrix of dgCMatrix. If It's a matrix, it should be integer. 

The error logs should point to the function and reason for users to debug. 

@igordot

**Review by evanbiederstedt — COMMENTED** [review]
Questions above @igordot

File: R/utils.R (hunk 1 above)
Comment (igordot) [reply R/utils.R:1638]: >  If It's a matrix, it should be integer.

A matrix can also be a `double`. For example, see `typeof(0)` vs `typeof(0L)`.

**Comment by evanbiederstedt** [comment]
@igordot 

> R uses mode "numeric" to mean double or integer.

I realize that :)

I'm asking to clarify if the error is only for integers, or for all "numeric" data.

**Comment by igordot** [comment]
> I'm asking to clarify if the error is only for integers, or for all "numeric" data.

I checked again. This was the original error:

```
running hclust...
Iteration 1
Mem used: 13.3Gb
Running HMMs on 3 cell groups..
Error in dpoilog(Y_obs, mu + log(phi * d * lambda_ref), sig) : 
  all x must be integers
```

I assumed that meant R integers, but it looks like they mean mathematical integers.

To reproduce, I tried slightly modifying the example dataset. This causes an error:

```
count_mat_ATC2_test = count_mat_ATC2
count_mat_ATC2_test = count_mat_ATC2_test + 0.001
```

This also causes an error:

```
count_mat_ATC2_test = count_mat_ATC2
count_mat_ATC2_test = count_mat_ATC2_test + 0.001
count_mat_ATC2_test = count_mat_ATC2_test - 0.001
```

This is fine:

```
count_mat_ATC2_test = count_mat_ATC2
count_mat_ATC2_test = count_mat_ATC2_test + 0.001
count_mat_ATC2_test = round(count_mat_ATC2_test, 0)
```

**Comment by evanbiederstedt** [comment]
RE: 

> Error in dpoilog(Y_obs, mu + log(phi * d * lambda_ref), sig) : 
  all x must be integers
  
 
 That's from here: https://github.com/kharchenkolab/numbat/blob/main/R/bipoilog.R#L86-L102
  
 That's taken directly from here: https://github.com/cran/poilog/blob/master/R/bipoilog.R
  
  
  ```
     dpoilog <- function(n,mu,sig){
     if (length(mu)>1 | length(sig)>1) stop('vectorization of mu and sig is currently not implemented') 
     if (any((n[n!=0]/trunc(n[n!=0]))!=1)) stop('all n must be integers')
     if (any(n<0)) stop('one or several values of n are negative')
     if (!all(is.finite(c(mu,sig)))) stop('all parameters should be finite')
     if (sig<=0) stop('sig is not larger than 0')
     .C('poilog1',n=as.integer(n),mu=as.double(mu),sig2=as.double(sig^2),nrN=as.integer(length(n)),val=double(length(n)))$val
   }
    
   ```
   
   And then note the function in C here: 
   
   https://github.com/cran/poilog/blob/master/src/bipoilog_s_cint.c#L127-L135
   
   Because of the interface with C, it has to be an integer. 
   
   
RE: `is.integer()`

This function doesn't work as you'd expect. 

```
> x = 5
> is.integer(x)
[1] FALSE
```

That's why I do the following:
```
## assume parameter 'foobar' must be integer

if (!is.numeric(foobar)) {
	stop("The parameter 'foobar' must be an integer. Please fix.")
}
if (foobar != as.integer(foobar)) {
	stop("The parameter 'foobar' must be an integer. Please fix.")
}
```

**Comment by igordot** [comment]
I changed it to `if (all(count_mat == round(count_mat, 0)))` in the latest version which seems to work for both standard and sparse matrices.

Anyway, I just wanted to move the check to the beginning. Otherwise it takes a while to see the error. Up to you how (and if) you want to implement it.

**Comment by evanbiederstedt** [comment]
Thanks, Igor

I'm not sure if there's a performance difference between the two solutions; I'll let @teng-gao check

This solution

```
if (all(count_mat == round(count_mat, 0))) {
    log_err("count_mat is not integers")
}
```

is the same as this:

```
if (!is.numeric(count_mat)) {
    log_err("count_mat must be of type 'integer'")
}
if (count_matr != as.integer(count_matr)) {
    log_err("count_mat must be of type 'integer'")
}
```

You can check this here:

```
ex1 <- matrix(c(3:14), nrow = 4, byrow = TRUE)
ex2 <-matrix(c(2.3, 3.4, 3.2, 1.1, 1.1, 5.5, 5.5, 6.5, 8.7), nrow = 3, byrow = TRUE)
```

That said, I would still do a `is.numeric()` check; you'll see `all(count_mat == round(count_mat, 0))` will throw obscure errors if the data type is totally different. 

Secondly, we should have a better error message for users to debug this themselves (as opposed to making GitHub issues)

**Comment by igordot** [comment]
> That said, I would still do a `is.numeric()` check

I think that would have a problem with a sparse matrix.

> Secondly, we should have a better error message for users to debug this themselves (as opposed to making GitHub issues)

I agree.

**Comment by evanbiederstedt** [comment]
> I think that would have a problem with a sparse matrix.

The check isn't for `dgCMatrix` but for matrices type `matrix`

The numeric check takes place after 

```
 if ('matrix' %in% class(count_mat)) 
 ```

**Comment by evanbiederstedt** [comment]
Here's my take on this @igordot 

https://github.com/kharchenkolab/numbat/pull/39

Please let me know if it works/what you think, etc.

**Comment by evanbiederstedt** [comment]
Thanks Igor!

Settled here:
https://github.com/kharchenkolab/numbat/pull/39

**Pull request closed without merging**